lib/dao.js

```diff
@@ -205,9 +205,11 @@ DataAccessObject.create = function (data, options, cb) {
         }
         data[i] = results[i].result;
       }
-      cb(errors, data);
+      if (cb.promise) {
+        cb(null, data);
+      } else cb(errors, data);
```

Comment:
This was the "partial success" case I mentioned...  At the time I just
decided to always resolve the promise, and then you would need to check the
error property of each result (which is what happens in the unit test).  I
don't really like it either...

I hadn't gotten around to changing the code since your last comment, but I
think you were correct in saying to drop promise support for the array
version.  (It shouldn't break existing code)

Actually, I think array creation with promises would be better handled at a
higher level, where the application developer can decide on the behavior.
e.g.:  with bluebird, you could do something like either
Promise.all(myArray).map(MyModel.create).all() or
Promise.all(myArray).map(MyModel.create).settled(), depending on your
requirements...

On Thu, Feb 19, 2015 at 9:59 AM, Miroslav Bajtoš <notifications@github.com>
wrote:

> In lib/dao.js
> <https://github.com/strongloop/loopback-datasource-juggler/pull/451#discussion_r25004186>
> :
>
> > @@ -205,9 +205,11 @@ DataAccessObject.create = function (data, options, cb) {
> >          }
> >          data[i] = results[i].result;
> >        }
> > -      cb(errors, data);
> > +      if (cb.promise) {
> > +        cb(null, data);
> > +      } else cb(errors, data);
>
> This does not look right, you are basically swallowing errors in the
> promise case.
>
> In the original PR, I was suggesting to drop support for the array version
> of create in the promise mode. Later, I have realised that such change
> can break the existing code calling the array version of create with no
> callback.
>
> Here is a better solution: when in the array and promise mode, build a
> single array where each item is either the model instance created or the
> error object when create failed. If there was at least one error, then
> reject the promise. If all models were created successfully, then resolve
> the promise with the array.
>
> Please make sure this case is covered by unit-tests (I did not check the
> test code yet).
>
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/strongloop/loopback-datasource-juggler/pull/451/files#r25004186>
> .
>